Write the 1.4 release notes (#605)
cert-manager/website: content/en/docs/release-notes/release-notes-1.4.md

* [@inteon](https://github.com/inteon)
* [@anton-johansson](https://github.com/anton-johansson)
* [@edglynes](https://github.com/edglynes)
* [@jandersen-plaid](https://github.com/jandersen-plaid)
* [@foosinn](https://github.com/foosinn)
* [@jsoref](https://github.com/jsoref)
* [@clatour](https://github.com/clatour)
* [@tamalsaha](https://github.com/tamalsaha)

# Deprecated Features and Breaking Changes

## Certificate renewal period calculation

There have been slight changes in the renewal period calculation for
certificates. Renewal time (`rt`) is now calculated using formula `rt = notAfter -
rb` where `rb = min(renewBefore, cert duration / 3)`. (See [docs](../../usage/certificate/#renewal) for more
detailed explanation). Previously this was calculated using formula `rt =
notAfter - rb`  where  `if cert duration < renewBefore; then rb = cert duration
/ 3; else rb = renewBefore`. This change fixes a bug where, if a certificate's
duration is very slightly larger than `renewBefore` period, then a cert gets
renewed at `notAfter - renewBefore` which can lead to a continuous renewal loop,
see [`cert-manager#3897`](https://github.com/jetstack/cert-manager/issues/3897).

> 

Author: I moved this to the "Bug Fixes" section (first entry) since I think it is now a breaking change anymore... or is it still?

@irbekrm

Reviewer: It does change the renewal behaviour if your certs have a duration of more than 90 days and `renewBefore` has not been set. Previously in that case the certs would have been renewed 30 days before expiry whereas now it will be 2/3 through the duration. Do you think that's worth being added as a change?

Author: Yes, I think that's a notable (and maybe breaking) change.

I moved the bug fix from under "Bug Fixes" to "Breaking Changes" in 119faf9 with a way to keep the old behavior by setting renewBefore to `720h`.

What do you think?

Reviewer: This looks great to me!

- `acme.cert-manager.io/v1alpha3`
- `acme.cert-manager.io/v1beta1`

These APIs will be removed in cert-manager 1.6.

{{% pageinfo color="warning" %}}

⛔️  If you have a cert-manager installation that is using or has previously used
these deprecated APIs, you may need to upgrade your cert-manager custom
resources and CRDs.

This needs to be done before upgrading to cert-manager 1.6. See cert-manager
[docs](../../installation/upgrading/remove-deprecated-apis/#upgrading-existing-cert-manager-resources)

> for more detailed upgrade instructions.

Reviewer 2: Can we be more specific about which users are affected by this? 

> If you are upgrading cert-manager on a cluster which has previously had cert-manager < v1.0.0 you MUST take steps to upgrade any cert-manager custom resources which were created by you or by cert-manager < v1.0.0.
> Such custom resources may still be stored in Etcd as v1alpha2 resources and when that version of the API is removed in cert-manager 1.6, they will become unreadable. 
> To work around this, you need to force them to be converted by the  cert-manager conversion webhook.
> And the simplest way to do that is to simply perform a no-change update on all cert-manager customer resources. 
> This is explained in more detail in the [Upgrading existing cert-manager resources] page.

Author: Thank you for this great suggestion, I added it verbatim in 647c958.

{{% /pageinfo %}}

## Helm chart: `securityContext` defaults to non-root

The Helm chart [now](https://github.com/jetstack/cert-manager/pull/4036) follows
the current Pod hardening best practices as defined by the Kyverno [pod-security

> restricted](https://kyverno.io/policies/pod-security/#restricted) policy.

Reviewer 2: I can't stand links like `now` and `docs` and `here and `this`. 
I suggest not linking to the PR.
```suggestion
The Helm chart now follows
the [Kyverno restricted Pod security policy](https://kyverno.io/policies/pod-security/#restricted).
```

To pass the validation, the controller, webbook, and cainjector Pods are now
running as non-root:

```yaml
apiVersion: v1
kind: Pod
spec:
  securityContext:
    runAsNonRoot: true
```

{{% pageinfo color="warning" %}}

⛔️  If you are using custom containers that run as root with the Helm chart, you
will need to set this back to `false`.

{{% /pageinfo %}}

## CA Issuer and `ca.crt`

The CA issuer [now](https://github.com/jetstack/cert-manager/pull/3865) attempts
to store the root CA instead of the issuing CA into the `ca.crt` field for
issued certificates. All of the information which was previously available is
still available: the intermediate should appear as part of the chain in

> `tls.crt`.

Reviewer 2: I suggest removing the `now` PR link.
And if you want to add PR links, do it at the end of each note as `Implemented in cert-manager PR#3865` (or something like that)

{{% pageinfo color="warning" %}}

> ⛔️  This is a change of behavior with the CA Issuer, make sure that the change in
> `ca.crt` does not impact you current use.

Reviewer 2: ```suggestion
⛔️  You may need to adjust systems that consume the `ca.crt` from Secrets managed by cert-manager with the CA issuer.
```

cert-manager. The CA Issuer is currently the only supported issuer. The feature
is experimental and can be enabled by adding a flag to the cert-manager
controller. For example, with Helm:

```sh
helm install cert-manager jetstack/cert-manager \
  --set extraArgs="{--feature-gates=ExperimentalCertificateSigningRequestControllers=true}"
```

Imagining that you have an existing CA issuer named `my-ca-issuer` in the
default namespace. To let cert-manager sign your CSR, you (the entity creating
the CSR) need a [specific
role](../../usage/kube-csr/#referencing-namespaced-issuers). You can then
create a CertificateSigningRequest:

```sh
openssl genrsa -out test.key 2048
kubectl apply -f - <<EOF
apiVersion: certificates.k8s.io/v1
kind: CertificateSigningRequest
metadata:
  name: test
spec:
  groups:
  - system:authenticated
  request: $(openssl req -new -key test.key -out /dev/stdout | base64 | tr -d "\n")
  signerName: issuers.cert-manager.io/default.my-ca-issuer
  usages:
  - client auth
EOF
```

You will still need to manually approve the CSR since cert-manager does not
approve CSRs by default
([unlike](../../concepts/certificaterequest/#approval)

> CertificateRequests that get automatically approved by default):

Reviewer 2: This seems like too much documentation to include in the release-notes. It should be in the kube-csr docs which we link to.

Author: I agree, it kind of duplicates the content of usage/kube-csr. I removed most of it in b0085737aa3e2803d2a18ebe871165d5b2c1572b.

```sh
kubectl certificate approve test
```

cert-manager then signs the CSR. To know more about it, see [the
documentation](../../usage/kube-csr) on how use the built-in
CertificateSigningRequests with cert-manager.

[CertificateSigningRequest]: https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/

## Vault Issuer handle of `ca.crt`

The Vault issuer is now able to fill the `ca.crt` in Secrets. To do that,
cert-manager reconstructs the certificate chain that is returned by Vault and
uses the certificate at the start of the chain as the `ca.crt`.

> 

Reviewer 2: "start of the chain" is ambiguous. 
If we are going to explain it, it should be in the docs for the Vault issuer.

Author: I think it is not worth detailing this, so I removed the mention to "start of the chain" in a3d6a65.

Note that this `ca.crt` may be an intermediate signing CA instead of the actual
root CA.

The reason why this feature did not exist previously is that finding which
certificate is at the top of the chain is not trivial. Take a look at
[@JoshVanL](https://github.com/JoshVanL)'s excellent
[`ParseSingleCertificateChain`](https://github.com/jetstack/cert-manager/blob/5e2a6883c1202739902ac94b5f4884152b810925/pkg/util/pki/parse.go#L167-L244)

> to see what this is all about!

Reviewer 2: I like that you're crediting Josh's code here, but it seems inappropriate for the release notes.
After all, you might argue that we should have been doing this from the start and we should be applying this cert chain parsing to the CertificateRequest CA information returned by all the issuers. 

And the fact that we're saying here that the CA may contain an intermediate just adds to the general confusion around `ca.crt` and contradicts the CA bug fix described above where we say that the `ca.crt` always contains a root crt.

Author: I removed this part in 827ec9e

Author: After [a discussion](https://kubernetes.slack.com/archives/CDEQJ0Q8M/p1623770041428700) with Josh, I rephrased the paragraph in a1744c9. Can you take a look?

### Helm chart: webhook externally accessible for bare-metal

In [some](https://github.com/kubernetes/kubernetes/issues/72936#issue-399522387)
Kubernetes setups, the apiserver is not able to talk to kube-dns (i.e., when
Kubernetes is running on bare-metal with no special `resolv.conf`).

To work around that, the cert-manager webhook can
[now](https://github.com/jetstack/cert-manager/pull/3876) be configured to be
accessible from outside of the cluster. For example, in `values.yaml`:

```yaml
# values.yaml
webhook:
  serviceType: LoadBalancer
  loadBalancerIP: 85.78.90.100
  url:
    host: 85.78.90.100

> ```

Reviewer 2: Again, I suggest updating the webhook docs and linking to that.

  keys.
- The upgrade from v1.3.0 to 1.4.0 with Helm and `--set installCRDs=true` is
  [now](https://github.com/jetstack/cert-manager/pull/3882) possible. This issue
  was due to [a Helm
  bug](https://github.com/helm/helm/issues/5806#issuecomment-788116838) with the
  `minimum` field on the CRDs.
- The ACME issuer [now](https://github.com/jetstack/cert-manager/pull/3866)
  works correctly with Certificates that have a long name (52 characters or
  more). These Certificates would not get renewed due to non-unique `Order`
  names being generated.
- Orders that are used with a misbehaving ACME server should not get stuck
  [anymore](https://github.com/jetstack/cert-manager/pull/3805). By misbehaving,
  we mean an ACME server that would validate the authorizations before having
  set the status of the order to "ready".
- The internal signers [now](https://github.com/jetstack/cert-manager/pull/3878)
  set the condition `Ready=False` with the reason `RequestDenied` when a
  CertificateRequest is `Denied`. This is to keep the same behavior where a
  terminal state of a CertificateRequest should have a `Ready` condition.

# Honorable mentions

> 

Reviewer 2: I know I'm too straight laced, but I just don't think this is appropriate for the release notes. It would be appropriate for a blog post.
Mentioning a not-merged feature in the release notes is just going to confuse people I think.

```yaml
apiVersion: cert-manager.io/v1
kind: Issuer
metadata:
  namespace: default
  name: my-ca-issuer
spec:
  ca:
    secretName: ca-key-pair
```

You will make sure to be given the following role so that cert-manager can sign
your CSR:

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
rules:
- apiGroups:
  - cert-manager.io
  resources:
  - signers
  verbs:
  - reference
  namespaces:
  - default
  resourceNames:
  - my-ca-issuer
```

You can create your Kubernetes CSR:

```sh
openssl genrsa -out test.key 2048
kubectl apply -f - <<EOF
apiVersion: certificates.k8s.io/v1
kind: CertificateSigningRequest
metadata:
  name: test
spec:
  groups:
  - system:authenticated
  request: $(openssl req -new -key test.key -out /dev/stdout | base64 | tr -d "\n")
  signerName: kubernetes.io/kube-apiserver-client
  usages:
  - client auth
EOF
```

You will still need to manually approve the CSR since cert-manager does not
approve CSRs by default
([unlike](../../concepts/certificaterequest/#approval)
CertificateRequests that get automatically approved by default):

```sh
kubectl certificate approve test
```

See [the documentation](../../usage/kube-csr) on how use the built-in

> CertificateSigningRequests with cert-manager.

Author: this link fails since it only exists in next-docs right now. When release-next is merged into master, the issue will go away. I don't know how to work around Prow regarding that though

Should I use the link 

```md
[the documentation](/next-docs/usage/kube-csr)
```

instead?

Reviewer: > When release-next is merged into master

Since we are releasing today, do you think we can just somehow co-ordinate the merging so that we merge release-next into master before merging this PR?

Reviewer: Also looking at the error message, might it just be the trailing slash?:

```
docs/release-notes/release-notes-1.4/index.html
  target is a directory, href lacks trailing slash --- docs/release-notes/release-notes-1.4/index.html --> ../../usage/kube-csr
```

Author: Lol once again I did not read the error message and assumed it was a broken link. Thank you for taking a look at the Prow build logs 🙏

```

With the above example, the source label
`__meta_kubernetes_service_label_app='armada-api'` becomes the new label
`app='armada-api'` when metrics related to this Service are scraped.

## Akamai issuer and Open Edgegrid EdgeDNS v2

The Akamai issuer has been
[updated](https://github.com/jetstack/cert-manager/pull/4007) to use Open
Edgegrid EdgeDNS v2 API.

# Bug Fixes

- The default certificate renewal duration of 30 days was clashing with the
  duration of certificates issued by the Vault PKI. All Certificates [are
  now](https://github.com/jetstack/cert-manager/pull/4092) renewed 2/3 through
  the duration unless a custom renew period is specified with the setting
  `spec.renewBefore` on the Certificate.
- The [RFC2136](https://cert-manager.io/docs/configuration/acme/dns01/rfc2136/)
  issuer is [now](https://github.com/jetstack/cert-manager/pull/3622) able to
  handle DNS-01 challenges that map to multiple `TXT` records. This lets you
  create Let's Encrypt certificates using RFC2136 with multiple DNS names.
- The comparison function `PublicKeysEqual` is
  [now](https://github.com/jetstack/cert-manager/pull/3914) correct for public
  keys.

> - The upgrade from v1.3.0 to 1.4.0 with Helm and `--set installCRDs=true` is

Reviewer: This was done so it's possible to upgrade from `v1.2.x` to  `v1.3.x`. I think this might not be relevant for `v1.4`?

Author: Should I rephrase as

```suggestion
- The upgrade from 1.2 to 1.3 with Helm and `--set installCRDs=true` is
```

Reviewer: > I have tested v1.3.0 -> latest master (so with this PR) upgrade on kind clusters with k8s v1.20 and v1.19 (with helm)

The goal of the PR was to fix a bug where users couldn't upgrade _to_ `v1.3.x`. The other comments are mostly for edge cases. We also didn't merge this PR into `v1.3.0`.

Author: Wait, so this PR is only meant to fix e.g. `v1.2.0` -> `v1.3.1` but we did not backport it to `release-1.3`?

Author: I shall then create a new `1.3.1` entry in the 1.3 release notes page 😅

  bug](https://github.com/helm/helm/issues/5806#issuecomment-788116838) with the
  `minimum` field on the CRDs.
- The ACME issuer [now](https://github.com/jetstack/cert-manager/pull/3866)
  works correctly with Certificates that have a long name (52 characters or
  more). These Certificates would not get renewed due to non-unique `Order`
  names being generated.
- Orders that are used with a misbehaving ACME server should not get stuck
  [anymore](https://github.com/jetstack/cert-manager/pull/3805). By misbehaving,
  we mean an ACME server that would validate the authorizations before having
  set the status of the order to "ready".
- The internal signers [now](https://github.com/jetstack/cert-manager/pull/3878)
  set the condition `Ready=False` with the reason `RequestDenied` when a
  CertificateRequest is `Denied`. This is to keep the same behavior where a
  terminal state of a CertificateRequest should have a `Ready` condition.

# Honorable mentions

Tim Ramlot ([@inteon](https://github.com/inteon)) has done a fantastic job at
adding the Istio VirtualService support for HTTP01 challenges in
[#3724](https://github.com/jetstack/cert-manager/pull/3724). It took an immense

> effort to have this PR ready and merged for the 1.4 release.

Reviewer: Nit: I think this may have been for the 1.3 release?

Author: The revert took place in https://github.com/jetstack/cert-manager/commit/79d8d9cb7b5186f0fcffe2e24e1198656737f26d, so I assumed that this PR was meant to be merged as part of the 1.4 release?

Reviewer: You're right. We _wanted_ to get it into v1.3 release 😄

After a lot of thinking, we have decided that trying to support every custom
resource for every proxy could not be done in-tree due to the Go dependency
weight that each integration adds. Jake Sanders proposed an [out-of-tree
approach](https://github.com/jetstack/cert-manager/issues/3924), and we have Tim

> [has been

Reviewer: Suggestion: 'who has been accepted'

Author: Should I still apply this change? The sentence is now (a5a3b27):

Tim [has been
accepted](https://summerofcode.withgoogle.com/projects/#4841881566969856) for
the 2021 Google Summer of Code to work on the `kubectl cert-manager install`
command.

Reviewer: I think this is correct - @wallrj could confirm that?

Author: I'll click "Resolve conversation" for now, feel free to re-open if you think this "Honorable mentions" should not be there @wallrj

accepted](https://summerofcode.withgoogle.com/projects/#4841881566969856) for

> the 2021 Google Summer of Code to work on this feature!

Reviewer: I think Tim will work on install command for kubectl cert-manager plugin?

Author: I thought we had submitted a new proposal, it [looks like not](https://github.com/cncf/mentoring/pulls?q=cert-manager)... Sorry, I will rephrase that and say that he is working on the CLI stuff.

Reviewer: I think it _may_ be this issue that is related to his proposal https://github.com/jetstack/cert-manager/issues/3952

- `acme.cert-manager.io/v1alpha3`
- `acme.cert-manager.io/v1beta1`

These APIs will be removed in cert-manager 1.6.

{{% pageinfo color="warning" %}}

> ⛔️  If you have a cert-manager installation that is using or has previously used
> these deprecated APIs, you may need to upgrade your cert-manager custom
> resources and CRDs.
> 
> This needs to be done before upgrading to cert-manager 1.6. See cert-manager
> [docs](../../installation/upgrading/remove-deprecated-apis/#upgrading-existing-cert-manager-resources)
> for more detailed upgrade instructions.

Reviewer 3: suggestion: with this being in a warning which will draw peoples' attention, we should be really clear that this isn't an _immediate_ call to action.

```suggestion
⛔️  If you have a cert-manager installation that is using or has previously used
these deprecated APIs, you may need to upgrade your cert-manager custom
resources and CRDs before the release of cert-manager version 1.6.

See the ["remove deprecated APIs" docs](../../installation/upgrading/remove-deprecated-apis/#upgrading-existing-cert-manager-resources)
 for more detailed upgrade instructions.
```

Author: Richard suggested a new phrasing in 647c958 which removes the feeling of "it should be done immediately". Let me know if that's still feeling too much.

  --set extraArgs="{--feature-gates=ExperimentalCertificateSigningRequestControllers=true}"
```

Note that you will still need to manually approve the CSR object before
cert-manager can sign the CSR.

The documentation is available on the [the Kubernetes CSR usage
page](../../usage/kube-csr/).

[CertificateSigningRequest]: https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/

### Vault Issuer handle of `ca.crt`

The Vault issuer is now able to fill the `ca.crt` in Secrets. The `ca.crt` set
by cert-manager may be an intermediate signing CA instead of the actual root CA.

### Helm chart: webhook externally accessible for bare-metal

In [some](https://github.com/kubernetes/kubernetes/issues/72936#issue-399522387)
Kubernetes setups, the apiserver is not able to talk to `kube-dns` (i.e., when
Kubernetes is running on bare-metal with no special `resolv.conf`).

To work around that, the cert-manager webhook can now be configured to be
accessible from outside of the cluster. For example, in `values.yaml`:

```yaml
# values.yaml
webhook:
  serviceType: LoadBalancer
  loadBalancerIP: 85.78.90.100
  url:
    host: 85.78.90.100

> ```

Author: Thanks for the suggestion, I was not aware of "example IPv4 blocks"! 👍

### Helm chart: Service labels

The cert-manager controller Service now supports custom labels using the
top-level field in `values.yaml`:

```yaml
# values.yaml
serviceLabels:
  app: armada-api
```

This may be useful in conjunction with Prometheus' `labelmap`. For example, with
the following sample Prometheus configuration:

```yaml
# prometheus.yaml
- action: labelmap
  regex: __meta_kubernetes_service_label_(.+)
```

With the above example, the source label
`__meta_kubernetes_service_label_app='armada-api'` becomes the new label
`app='armada-api'` when metrics related to this Service are scraped.

### Akamai issuer

The Akamai issuer has been
[updated](https://github.com/jetstack/cert-manager/pull/4007) to use the v2 of

> the `OPEN EdgeGrid` Go package.

Author: Thanks! I also changed the title of the PR (issuer -> DNS01 solver)

spec:
  securityContext:
    runAsNonRoot: true
```

{{% pageinfo color="warning" %}}

⛔️  If you are using custom containers that run as root with the Helm chart, you
will need to set this back to `false`.

{{% /pageinfo %}}

### CA, Vault and Venafi issuer handling of `ca.crt` and `tls.crt`

The CA, Vault, and Venafi issuer now produce a `tls.crt` that is de-duplicated,
in the correct order (leaf at the top, issuing certificate at the bottom) and
verified (i.e. each signature can be verified).

The CA issuer now produces a `ca.crt` that contains the root-most CA instead of
the intermediate CA when the issuing certificate is an intermediate CA. By
root-most, we mean that `ca.crt` is the root of the chain of certificates that
cert-manager knows about, but it might not be the self-signed root CA.

[#3982]: https://github.com/jetstack/cert-manager/pull/3982 "All issuers + Vault issuer"
[#3983]: https://github.com/jetstack/cert-manager/pull/3983 "Venafi issuer"
[#3985]: https://github.com/jetstack/cert-manager/pull/3985 "CA issuer"

{{% pageinfo color="warning" %}}

⛔️  You may need to adjust systems that consume the `ca.crt` from Secrets
managed by cert-manager with the CA issuer.

> {{% /pageinfo %}}

Author: @erikgb @JoshVanL @wallrj does this paragraph make sense?

Context: [this Slack thread](https://kubernetes.slack.com/archives/CDEQJ0Q8M/p1623770041428700)

Reviewer 4: Makes sense to me :+1:

Reviewer 5: Makes sense to me too, but this is a definately bit "complex". The `ca.crt` will contain the "most" root CA it is aware of. 😄

Reviewer 5: I mean: The self-signed CA, aka root-CA is not really required to issue certificates.

Author: Thanks Erik!! Your phrasing is clearer. I went with:

> The CA issuer now produces a `ca.crt` that contains the "most" root CA that cert-manager is aware of. `ca.crt` may thus not be the actual self-signed root CA, since cert-manager may not be aware of it.

- `acme.cert-manager.io/v1alpha3`
- `acme.cert-manager.io/v1beta1`

These APIs will be removed in cert-manager 1.6.

{{% pageinfo color="warning" %}}

> ⛔️  If you are upgrading cert-manager on a cluster which has previously had
> cert-manager < `v1.0.0`, you **must** take steps to upgrade all cert-manager
> custom resources which were created by you or by cert-manager < `v1.0.0`.
> 
> Such custom resources may still be stored in `etcd` as `v1alpha2` resources and
> when that version of the API is removed in cert-manager 1.6, they will become
> unreadable.
> 
> To work around this, you need to force them to be converted by the cert-manager
> conversion webhook. And the simplest way to do that is to perform a no-change
> update on all cert-manager custom resources.
> 
> This is explained in more detail in the [Upgrading existing cert-manager
> resources][upgrade-resources] page.

Reviewer: ```suggestion
⛔️  If you are upgrading cert-manager on a cluster which has previously had
cert-manager < `v1.0.0`, you will need to ensure that all cert-manager custom resources
are stored in etcd at `v1` version and that cert-manger CRDs do not reference the deprecated APIs
**by the time you upgrade to `v1.6`**.

This is explained in more detail in the [Upgrading existing cert-manager
resources][upgrade-resources] page.
```

Reviewer: Let me know if this makes sense ^. I think it might make sense to give this a bit less urgency- users only need to do this by the time they upgrade to v1.6 not now, so it might be confusing.
